File: backend/app/routers/creator/campaigns.py
"""Creator-facing campaign browse + detail + join. Active campaigns only."""
from __future__ import annotations

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.core.deps import get_current_creator
from app.db.session import get_db
from app.models import Campaign, Creator
from app.schemas.campaign import BonusMilestoneOut, CampaignPublicOut, MyCampaignOut, ParticipationOut
from app.services import campaign as svc

router = APIRouter(prefix="/campaigns", tags=["creator-campaigns"])


@router.get("/mine", response_model=list[MyCampaignOut])
def mine(current: Creator = Depends(get_current_creator), db: Session = Depends(get_db)):
    """Campaigns the creator applied to / joined, with their application status."""
    return [MyCampaignOut(**m) for m in svc.list_creator_campaigns(db, current.id)]


def _public_out(c: Campaign, joined: bool, db: Session | None = None) -> CampaignPublicOut:
    milestones: list[BonusMilestoneOut] = []
    if db is not None:
        milestones = [
            BonusMilestoneOut(
                id=str(m.id), views_threshold=m.views_threshold, bonus_amount=m.bonus_amount,
                description=m.description, sort_order=m.sort_order,
            )
            for m in svc.get_bonus_milestones(db, c.id)
        ]
    return CampaignPublicOut(
        id=str(c.id), slug=c.slug, name=c.name, description=c.description, mode=c.mode,
        cpm_rate=c.cpm_rate, budget=c.budget, platforms=list(c.platforms or []),
        min_retention_days=c.min_retention_days, brief_script=c.brief_script,
        content_drive_url=c.content_drive_url, caption_rules=c.caption_rules,
        required_mentions=list(c.required_mentions or []), example_captions=list(c.example_captions or []),
        requirements_url=c.requirements_url, brand_name=c.brand_name, brand_logo_url=c.brand_logo_url,
        starts_at=c.starts_at, ends_at=c.ends_at, joined=joined,
        payment_type=c.payment_type, fixed_amount=c.fixed_amount,
        weekly_hours_needed=c.weekly_hours_needed, hourly_rate=c.hourly_rate,
        required_hours=c.required_hours, per_post_amount=c.per_post_amount,
        example_videos=list(c.example_videos or []), age_requirement=c.age_requirement,
        platform_focus=list(c.platform_focus or []), content_type=c.content_type,
        posting_frequency=c.posting_frequency, video_length=c.video_length,
        account_type=c.account_type, is_app=c.is_app, physical_product=c.physical_product,
        banner_url=c.banner_url, bonus_milestones=milestones,
        job_type=c.job_type, creator_type=c.creator_type,
    )


@router.get("", response_model=list[CampaignPublicOut])
def browse(current: Creator = Depends(get_current_creator), db: Session = Depends(get_db)):
    campaigns = svc.list_active_campaigns(db)
    return [_public_out(c, svc.creator_has_joined(db, c.id, current.id), db) for c in campaigns]


@router.get("/{slug}", response_model=CampaignPublicOut)
def detail(slug: str, current: Creator = Depends(get_current_creator), db: Session = Depends(get_db)):
    c = svc.get_active_campaign(db, slug)
    return _public_out(c, svc.creator_has_joined(db, c.id, current.id), db)


@router.post("/{slug}/join", response_model=ParticipationOut)
def join(slug: str, current: Creator = Depends(get_current_creator), db: Session = Depends(get_db)):
    part = svc.join_campaign(db, current.id, slug)
    return ParticipationOut(id=str(part.id), campaign_id=str(part.campaign_id), status=part.status)

File: frontend/src/lib/campaigns.ts
import { apiFetch, finalizeUpload, presignUpload, putToPresignedUrl, uploadCapturedPoster } from "@/lib/api";
import { getAuthToken } from "@/lib/auth";

const auth = () => ({ token: getAuthToken() ?? undefined });

// Bonus milestone row (Feature 3 wizard step 3) — surfaced natively on the
// campaign brief page (Feature 5) instead of living only in the admin builder.
export type BonusMilestone = {
  id: string;
  views_threshold: number;
  bonus_amount: number;
  description: string | null;
  sort_order: number;
};

// Wizard fields (Feature 3, BUILD_SPEC.md §3.3) shared by the creator-facing
// campaign types (CampaignPublicOut on the backend) — every field optional /
// defaulted so legacy campaigns created before the wizard still render fine.
export type CampaignWizardFields = {
  payment_type: string | null;
  fixed_amount: number | null;
  fixed_unit: string | null;
  weekly_hours_needed: number | null;
  hourly_rate: number | null;
  required_hours: number | null;
  per_post_amount: number | null;
  example_videos: string[];
  examples?: { url: string; platform?: string | null; thumbnail_url?: string | null }[];
  age_requirement: string | null;
  platform_focus: string[];
  content_type: string | null;
  posting_frequency: string | null;
  video_length: string | null;
  account_type: string | null;
  is_app: boolean;
  physical_product: boolean;
  banner_url: string | null;
  bonus_milestones: BonusMilestone[];
  job_type: string | null;
  creator_type: string | null;
};

export type Campaign = {
  id: string;
  slug: string;
  name: string;
  description: string | null;
  mode: "create_new" | "copy_paste";
  cpm_rate: number;
  budget: number;
  platforms: string[];
  min_retention_days: number;
  brief_script: string | null;
  content_drive_url: string | null;
  caption_rules: string | null;
  required_mentions: string[];
  example_captions: string[];
  requirements_url: string | null;
  brand_name: string | null;
  brand_logo_url: string | null;
  starts_at: string | null;
  ends_at: string | null;
  joined: boolean;
  approved: boolean;
} & CampaignWizardFields;

export type Participation = { id: string; campaign_id: string; status: string };
export type Submission = {
  id: string;
  campaign_id: string;
  post_url: string;
  platform: string;
  views: number;
  likes: number;
  comments: number;
  estimated_amount: number;
  payable_amount: number | null;
  scrape_status: string;
  verification_status: string;
  verification_note: string | null;
  revision_mode: "edit" | "repost" | null;
  has_proof_video: boolean;
  thumbnail_url: string | null;
  claimed: boolean;
  is_paid: boolean;
  created_at: string;
};

export const browseCampaigns = () => apiFetch<Campaign[]>("/api/creator/campaigns", auth());

// Campaigns an admin personally invited this creator to — same shape as browse,
// so they render with the identical Explore CampaignCard (thumbnail + Joined badge).
export const listInvitedCampaigns = () => apiFetch<Campaign[]>("/api/creator/campaigns/invited", auth());

export const listSubmissions = () =>
  apiFetch<Submission[]>("/api/creator/submissions", auth());

export const getSubmission = (id: string) =>
  apiFetch<Submission>(`/api/creator/submissions/${id}`, auth());

export const claimSubmission = (id: string) =>
  apiFetch<Submission>(`/api/creator/submissions/${id}/claim`, { method: "POST", ...auth() });

export const getCampaign = (slug: string) =>
  apiFetch<Campaign>(`/api/creator/campaigns/${slug}`, auth());

export const joinCampaign = (slug: string) =>
  apiFetch<Participation>(`/api/creator/campaigns/${slug}/join`, { method: "POST", ...auth() });

export const submitClip = (campaign_slug: string, post_url: string) =>
  apiFetch<Submission>("/api/creator/submissions", {
    method: "POST",
    body: JSON.stringify({ campaign_slug, post_url }),
    ...auth(),
  });

// Amend a revision-requested (mode 'edit') submission with a corrected link →
// back to pending review. campaign_slug is ignored server-side but keeps the
// SubmissionCreateIn shape.
export const resubmitClip = (id: string, post_url: string) =>
  apiFetch<Submission>(`/api/creator/submissions/${id}/resubmit`, {
    method: "POST",
    body: JSON.stringify({ campaign_slug: "", post_url }),
    ...auth(),
  });

// Uploads a proof-of-post video (presign -> PUT -> finalize) and links it to the
// submission — the create_new verification gate an admin reviews before payout.
export async function uploadProofVideo(
  submissionId: string,
  file: File,
  onProgress?: (pct: number) => void,
): Promise<Submission> {
  const token = getAuthToken() ?? undefined;
  const presigned = await presignUpload(token!, {
    purpose: "proof_video",
    content_type: file.type || undefined,
    filename: file.name || undefined,
    size_bytes: file.size,
  });
  // Capture the first frame in parallel with the video upload for an instant
  // thumbnail; best-effort, so a failure never blocks the proof submission.
  const [, thumbnail_url] = await Promise.all([
    putToPresignedUrl(presigned.upload_url, file, onProgress),
    uploadCapturedPoster(token!, file),
  ]);
  await finalizeUpload(token!, presigned.object_id);
  return apiFetch<Submission>(`/api/creator/submissions/${submissionId}/proof`, {
    method: "PATCH",
    body: JSON.stringify({ storage_object_id: presigned.object_id, thumbnail_url }),
    ...auth(),
  });
}
